### ISSUE TITLE
Create a methodology to test a trained agent

### ISSUE THREAD

**Issue by DanielCoelho112** [DanielCoelho112/curlad#8]

**Comment by DanielCoelho112** [comment]
To train an agent we should run the following command:

    python3 main.py --experiment_name dummy_test --visualize_monitor

To test, run:

    python3 main.py --experiment_name dummy_test -ttc test0.yaml --visualize_monitor

By default we are assuming we are training, if we want to test, we have to point to the testing config.

The structure of an experiment is the following:
![image](https://user-images.githubusercontent.com/73063947/199772137-35577d32-749b-4186-b0e0-d98dcb7cde12.png)

**Comment by miguelriemoliveira** [comment]
So testing is all the same as training, but without updating the network of the agent?

Which performance metrics will we have? Just the rewards?

**Comment by DanielCoelho112** [comment]
> So testing is all the same as training, but without updating the network of the agent?

In general terms yes. In test, we also load the agent's weights.

> Which performance metrics will we have? Just the rewards?

Usually, we don't consider the rewards as the performance metrics. Although in some cases makes sense to use them.

In deep learning, we usually have the loss function for training, for example cross entropy loss, and then some performance metrics, for example accuracy, recall, etc.
In reinforcement learning we have the reward for training and then some performance metrics. In this case, the performance metrics are described in #13 : [episode_stats.csv](https://github.com/DanielCoelho112/curlad/files/9910027/episode_stats.csv)

For instance, the score_penalty is computed using the following expression:

```python
  score_penalty = 1.0 * (1 - (outside_lane_length+wrong_lane_length)/completed_length) \
              * (PENALTY_COLLISION_STATIC ** n_collisions_layout) \
              * (PENALTY_COLLISION_VEHICLE ** n_collisions_vehicle) \
              * (PENALTY_COLLISION_PEDESTRIAN ** n_collisions_pedestrian) \
              * (PENALTY_TRAFFIC_LIGHT ** n_red_light) \
              * (PENALTY_STOP ** n_stop_infraction) \
```

**Comment by miguelriemoliveira** [comment]
> Usually, we don't consider the rewards as the performance metrics. Although in some cases makes sense to use them.

This is something that always puzzled me. Why not use the same cost function (or reward function) as the performance metrics by which the system will be evaluated in the end?

Its like studying just for the question you know will be in the exam, but it works.

I know the normal is to not have the cost function and the evaluation be the same, but the question is: why not?

**Comment by DanielCoelho112** [comment]
> I know the normal is to not have the cost function and the evaluation be the same, but the question is: why not?

Because the goal for these two functions is different.

The loss function is designed to help the algorithm converge into something that is aligned with the performance metric.
In other words, a good loss function is one that is easily optimizable.

By contrast, a performance metric tells the user how good the algorithm is. It should be something that has "real" meaning to the user, and not something that no one really understands but it is good for the optimization algorithm.

Taking for instance our problem. Let's suppose we want to use the following evaluation metric: Em = percentage_route_completed + (1 - alpha*n_collisions). 

This simplified evaluation metric is very good for us to understand if we have a good driving agent or not. However, it is not suitable for the reward function. It is too focused on the end goal for an agent to learn. What the majority of researchers use for the reward function is something more related to the immediate step, for example, horizontal distance w.r.t. next waypoint, or velocity towards the next waypoint.

Now you should be wondering: "Ok, but why not use the cumulative reward function as the evaluation metric?"
Because we don't care about the horizontal distance w.r.t. the next waypoint, we care if the agent collides with something and if it reaches the final destination. 

To conclude, the loss function/reward is designed for the agent to learn efficiently, and the evaluation metric is designed to convey to the user how good the algorithm is, based on what we really want.
To foster a good algorithm, the functions should share the same goal.

**Comment by miguelriemoliveira** [comment]
Ok, makes sense. Thanks for the detailed response.
